validator/db/kv/attestation_history_v2.go

```diff
@@ -127,6 +127,44 @@ func (hd EncHistoryData) SetTargetData(ctx context.Context, target uint64, histo
 	return hd, nil
 }
 
+// UpdateHistoryForAttestation returns a modified attestation history with specified target+epoch pairs marked
+// as attested for. This is used to prevent the validator client from signing any slashable attestations.
+func (hd EncHistoryData) UpdateHistoryForAttestation(
+	ctx context.Context,
+	sourceEpoch,
+	targetEpoch uint64,
+	signingRoot [32]byte,
+) (EncHistoryData, error) {
+	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
+	latestEpochWritten, err := hd.GetLatestEpochWritten(ctx)
+	if err != nil {
+		return EncHistoryData{}, errors.Wrap(err, "could not get latest epoch written from history")
+	}
+	currentHD := hd
+	if targetEpoch > latestEpochWritten {
+		// If the target epoch to mark is ahead of latest written epoch, override the old targets and mark the requested epoch.
+		// Limit the overwriting to one weak subjectivity period as further is not needed.
+		maxToWrite := latestEpochWritten + wsPeriod
+		for i := latestEpochWritten + 1; i < targetEpoch && i <= maxToWrite; i++ {
+			newHD, err := hd.SetTargetData(ctx, i%wsPeriod, &HistoryData{Source: params.BeaconConfig().FarFutureEpoch})
+			if err != nil {
+				return EncHistoryData{}, errors.Wrap(err, "could not set target data")
+			}
+			currentHD = newHD
+		}
+		newHD, err := currentHD.SetLatestEpochWritten(ctx, targetEpoch)
+		if err != nil {
+			return EncHistoryData{}, errors.Wrap(err, "could not set latest epoch written")
+		}
+		currentHD = newHD
+	}
+	newHD, err := currentHD.SetTargetData(ctx, targetEpoch%wsPeriod, &HistoryData{Source: sourceEpoch, SigningRoot: signingRoot[:]})
```

Comment:
Since github review, only allows me to comment on newly added in code I will add my thoughts here on. The way the type `EncHistoryData` has been used throughout this file and for local protection is  prone to misuse/bugs. This might have also lead to bugs in the past too, although I cannot confirm so far.

`EncHistoryData` is a byte slice aliased type through which we store our attestation histories. When we do want to store any particular attestation history we do `history.SetTargetData` . 

```go
func (hd EncHistoryData) SetTargetData(ctx context.Context, target uint64, historyData *HistoryData) (EncHistoryData, error) {
	if err := hd.assertSize(); err != nil {
		return nil, err
	}
	// Cursor for the location to write target epoch to.
	// Modulus of target epoch  X weak subjectivity period in order to have maximum size to the encapsulated data array.
	cursor := latestEpochWrittenSize + (target%params.BeaconConfig().WeakSubjectivityPeriod)*historySize
	if uint64(len(hd)) < cursor+historySize {
		ext := make([]byte, cursor+historySize-uint64(len(hd)))
		hd = append(hd, ext...)
	}
	copy(hd[cursor:cursor+sourceSize], bytesutil.Uint64ToBytesLittleEndian(historyData.Source))
	copy(hd[cursor+sourceSize:cursor+sourceSize+signingRootSize], historyData.SigningRoot)
	return hd, nil
}
```

Here is the relevant receiver method, the important thing to notice is that for all receiver methods in`EncHistoryData we reference the value of the object rather than its pointer ex:
```
func (hd EncHistoryData)
```
This isn't an issue if the intention of the method is simply to read from the byte array as the backing array the slice references stays the same. However the moment you do want to write to the array,etc it becomes tricky and prone
to unintended side effects. Ex:

```go

	if uint64(len(hd)) < cursor+historySize {
		ext := make([]byte, cursor+historySize-uint64(len(hd)))
		hd = append(hd, ext...)
	}
```
This chunk is from the above linked method, we expand the slice here, however given the receiver method only takes the value instead of reference, the expanded slice isn't being referenced anywhere except in the method.  To illustrate this:

https://play.golang.org/p/lx8VF63Jxk6

The actual slice hasn't been expanded. Due to this fact, we always have to return the actual attestation history in the method. So that the slice object we hold is always 'correct' . 

so we do things like 
```
		currentHD = newHD
```
multiple times in a method. Doing it this way is bug prone as anyone writing the code will always have to remember to 
set the value this way despite using a receiver method that says:
```go
obj.SetTargetData
```
For most people, they would assume that the actual object is changed but in reality only a copy of it has been modified.

It will be worth it to audit how exactly attestation histories have been used throughout the repo, and that we haven't used it incorrectly and set the wrong history,etc.